tests/integration/upgrade_test.go

```diff
@@ -0,0 +1,132 @@
+	s.helper = clients.CreateTestClient(s.k8sh, s.op.installer.Manifests)
+}
+
+func (s *UpgradeSuite) TearDownSuite() {
+	s.op.Teardown()
+}
+
+func (s *UpgradeSuite) TestUpgradeToMaster() {
+	systemNamespace := installer.SystemNamespace(s.namespace)
+
+	// Create block, object, and file storage on 0.8 before the upgrade
+	poolName := "upgradepool"
+	storageClassName := "block-upgrade"
+	blockName := "block-claim-upgrade"
+	podName := "test-pod-upgrade"
+	logger.Infof("Initializing block before the upgrade")
+	setupBlockLite(s.helper, s.k8sh, s.Suite, s.namespace, poolName, storageClassName, blockName, podName)
+	createPodWithBlock(s.helper, s.k8sh, s.Suite, s.namespace, blockName, podName)
+	defer blockTestDataCleanUp(s.helper, s.k8sh, s.namespace, poolName, storageClassName, blockName, podName)
+
+	logger.Infof("Initializing file before the upgrade")
+	filesystemName := "upgrade-test-fs"
+	createFilesystem(s.helper, s.k8sh, s.Suite, s.namespace, filesystemName)
+	createFilesystemConsumerPod(s.helper, s.k8sh, s.Suite, s.namespace, filesystemName)
+	defer cleanupFilesystemConsumer(s.helper, s.k8sh, s.Suite, s.namespace, filesystemName)
+
+	logger.Infof("Initializing object before the upgrade")
+	objectStoreName := "upgraded-object"
+	runObjectE2ETestLite(s.helper, s.k8sh, s.Suite, s.namespace, objectStoreName, 1)
+
+	// verify that we're actually running 0.8 before the upgrade
+	operatorContainer := "rook-ceph-operator"
+	version, err := k8sutil.GetDeploymentVersion(s.k8sh.Clientset, systemNamespace, operatorContainer, operatorContainer)
+	assert.Nil(s.T(), err)
+	assert.Equal(s.T(), "v0.8.1", version)
+
+	s.simpleWriteAndRead(podName, "pre-upgrade-file")
+
+	// Upgrade to master
+	_, err = s.k8sh.Kubectl("-n", systemNamespace, "set", "image", "deploy/"+operatorContainer, operatorContainer+"=rook/ceph:"+installer.VersionMaster)
+	require.Nil(s.T(), err)
+
+	// verify that the operator spec is updated
+	version, err = k8sutil.GetDeploymentVersion(s.k8sh.Clientset, systemNamespace, operatorContainer, operatorContainer)
+	assert.Nil(s.T(), err)
+	assert.Equal(s.T(), installer.VersionMaster, version)
+
+	// wait for the osd pods to be updated
+	err = k8sutil.WaitForDeploymentVersion(s.k8sh.Clientset, s.namespace, "app=rook-ceph-osd", "rook-ceph-osd", installer.VersionMaster)
+	require.Nil(s.T(), err)
+	logger.Infof("Done with automatic upgrade to master")
+
+	// test writing and reading from the mounts that were created before the upgrade
+	s.simpleWriteAndRead(podName, "post-upgrade-file")
+}
+
+func (s *UpgradeSuite) simpleWriteAndRead(podName, filename string) {
+	// Verify the block storage is functional
+	logger.Infof("Write to block storage")
+	message := "basic data written after upgrade"
+	_, wtErr := s.helper.BlockClient.Write(podName, message, filename, "")
+	require.Nil(s.T(), wtErr)
+
+	logger.Infof("Read from block storage")
+	data, rErr := s.helper.BlockClient.Read(podName, filename, "")
```

Comment:
looks like we are just reading from the file that we just wrote.  i'd say that it would be interesting to *after* the upgrade also read the file that was written *before* the upgrade, ensuring that the OSDs are fully healthy and PGs/data/objects that existed before the upgrade are still in good condition.

Reply:
agreed, will add a read to the original file